internal/app/machined/pkg/system/services/extension.go

```diff
@@ -107,7 +107,7 @@ func (svc *Extension) getOCIOptions(envVars []string) []oci.SpecOpts {
 	ociOpts := []oci.SpecOpts{
 		oci.WithRootFSPath(filepath.Join(constants.ExtensionServicesRootfsPath, svc.Spec.Name)),
 		containerd.WithRootfsPropagation(svc.Spec.Container.Security.RootfsPropagation),
-		oci.WithCgroup(constants.CgroupExtensions),
+		oci.WithCgroup(filepath.Join(constants.CgroupExtensions, svc.Spec.Name)),
```

Comment:
this is the real fix, everything else is more to stress that service spec is not shared between services, so we'd rather not pass it by reference